File: packages/nodejs/ios/Plugin/NodejsPlugin.swift
import Foundation
import Capacitor

@objc(NodejsPlugin)
public class NodejsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NodejsPlugin"
    public let jsName = "Nodejs"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isReady", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "send", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise)
    ]

    public static let eventMessage = "message"
    public static let eventReady = "ready"
    public static let startModeAuto = "auto"
    public static let startModeManual = "manual"
    public static let tag = "NodejsPlugin"

    private var config = NodejsConfig()
    private var implementation: Nodejs?

    override public func load() {
        config = getNodejsConfig()
        implementation = Nodejs(plugin: self, config: config)
        if config.startMode == NodejsPlugin.startModeAuto {
            implementation?.start(StartOptions()) { error in
                if let error = error {
                    CAPLog.print("[", NodejsPlugin.tag, "] ", error)
                }
            }
        }
    }

    @objc func isReady(_ call: CAPPluginCall) {
        let result = IsReadyResult(ready: implementation?.isReady() == true)
        resolveCall(call, result)
    }

    @objc func send(_ call: CAPPluginCall) {
        do {
            let options = try SendOptions(call)

            try implementation?.send(options)
            resolveCall(call)
        } catch {
            rejectCall(call, error)
        }
    }

    @objc func start(_ call: CAPPluginCall) {
        guard config.startMode == NodejsPlugin.startModeManual else {
            rejectCall(call, CustomError.startModeNotManual)
            return
        }
        let options = StartOptions(call)

        implementation?.start(options) { error in
            if let error = error {
                self.rejectCall(call, error)
            } else {
                self.resolveCall(call)
            }
        }
    }

    func notifyMessageListeners(_ event: MessageEvent) {
        notifyListeners(NodejsPlugin.eventMessage, data: event.toJSObject() as? JSObject ?? JSObject(), retainUntilConsumed: true)
    }

    func notifyReadyListeners() {
        notifyListeners(NodejsPlugin.eventReady, data: JSObject(), retainUntilConsumed: true)
    }

    private func getNodejsConfig() -> NodejsConfig {
        var config = NodejsConfig()
        if let nodeDir = getConfig().getString("nodeDir") {
            config.nodeDir = nodeDir
        }
        if let startMode = getConfig().getString("startMode") {
            config.startMode = startMode
        }
        return config
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", NodejsPlugin.tag, "] ", error)
        call.reject(error.localizedDescription, (error as? CustomError)?.code)
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/nodejs/android/src/main/java/io/capawesome/capacitorjs/plugins/nodejs/NodejsPlugin.java
package io.capawesome.capacitorjs.plugins.nodejs;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.nodejs.classes.CustomException;
import io.capawesome.capacitorjs.plugins.nodejs.classes.CustomExceptions;
import io.capawesome.capacitorjs.plugins.nodejs.classes.events.MessageEvent;
import io.capawesome.capacitorjs.plugins.nodejs.classes.options.SendOptions;
import io.capawesome.capacitorjs.plugins.nodejs.classes.options.StartOptions;
import io.capawesome.capacitorjs.plugins.nodejs.classes.results.IsReadyResult;
import io.capawesome.capacitorjs.plugins.nodejs.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.nodejs.interfaces.Result;

@CapacitorPlugin(name = "Nodejs")
public class NodejsPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String EVENT_MESSAGE = "message";
    public static final String EVENT_READY = "ready";
    public static final String START_MODE_AUTO = "auto";
    public static final String START_MODE_MANUAL = "manual";
    public static final String TAG = "NodejsPlugin";

    private NodejsConfig config;
    private Nodejs implementation;

    @Override
    public void load() {
        config = getNodejsConfig();
        implementation = new Nodejs(this, getContext(), config);
        if (START_MODE_AUTO.equals(config.getStartMode())) {
            implementation.start(
                new StartOptions(),
                new EmptyCallback() {
                    @Override
                    public void success() {}

                    @Override
                    public void error(Exception exception) {
                        Logger.error(TAG, "Failed to start the Node.js runtime.", exception);
                    }
                }
            );
        }
    }

    @PluginMethod
    public void isReady(PluginCall call) {
        try {
            IsReadyResult result = new IsReadyResult(implementation.isReady());
            resolveCall(call, result);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void send(PluginCall call) {
        try {
            SendOptions options = new SendOptions(call);

            implementation.send(options);
            resolveCall(call);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void start(PluginCall call) {
        try {
            if (!START_MODE_MANUAL.equals(config.getStartMode())) {
                rejectCall(call, CustomExceptions.START_MODE_NOT_MANUAL);
                return;
            }
            StartOptions options = new StartOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.start(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    public void notifyMessageListeners(@NonNull MessageEvent event) {
        notifyListeners(EVENT_MESSAGE, event.toJSObject(), true);
    }

    public void notifyReadyListeners() {
        notifyListeners(EVENT_READY, new JSObject(), true);
    }

    @Override
    protected void handleOnPause() {
        super.handleOnPause();
        implementation.handleOnPause();
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        implementation.handleOnResume();
    }

    @NonNull
    private NodejsConfig getNodejsConfig() {
        NodejsConfig config = new NodejsConfig();
        String nodeDir = getConfig().getString("nodeDir", null);
        if (nodeDir != null) {
            config.setNodeDir(nodeDir);
        }
        String startMode = getConfig().getString("startMode", null);
        if (startMode != null) {
            config.setStartMode(startMode);
        }
        return config;
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
